conjure-java-core/src/integrationInput/java/com/palantir/product/EmptyPathServiceEndpoints.java

```diff
@@ -23,11 +22,11 @@ public static Service of(UndertowEmptyPathService delegate) {
     }
 
     @Override
-    public Registrable create(ServiceContext context) {
-        return new EmptyPathServiceRegistrable(context, delegate);
+    public void register(ServiceContext context, EndpointRegistry registry) {
```

Comment:
what's the rationale behind passing in an EndpointRegistry instead of returning an EndpointRegistry?

Reply:
Many services are registered to a single EndpointRegistry, and we have no guarantee that each service has a unique prefix.

Reply:
so an alternative approach would be returning an "EndpointList" or having the ability to merge many EndpointRegistry objects. Why do the conditions you mention demand that we invert the control here?

I think, generally, I'd prefer to see us use a more declarative approach, rather than passing in an object that gets configured, we ask this class to produce required metadata and then something external to it uses that information to do the configuring.

Reply:
That is exactly what `Registerable` did. Perhaps renaming `Registerable` to `EndpointList` would be better, though you noted that you didn't like the double-initialization of the declarative model before.

Reply:
Actually it was the same objection: Registrable returned a thing that accepts an EndpointRegistry and performs registration of endpoints. I'd like to return a thing that that an EndpointRegistry would iterate over to perform endpoint registration.

```
interface Registrable {
    Iterable<EndpointDefinition> endpoints();
}

interface EndpointDefinition {
    Endpoint endpoint();
    Handler handler();
}
```

Or, we could simplify a bit:

```
interface Service {
        Iterable<EndpointDefinition> endpoints(ServiceContext context);
}
```

Reply:
Ack, I will make the second change.

Reply:
@rubenfiszel Thoughts on including `HttpHandler` in `Endpoint` to avoid the `EndpointDefinition` tuple type?

Reply:
@cakofony alternatively perhaps it just returns a `Map<Endpoint, HttpHandler>`?

Reply:
I'm not a huge fan of using maps in the API -- that could silently lose data if we attempt to combine two results, where duplicate Endpoints in a list would fail upon registration to the registry, which is cleaner behavior.

Currently our `Endpoint` is just the metadata describing where the endpoint can be reached, the `HttpHandler` _is_ the endpoint. The more I think about this the better I like combining the handler into our `Endpoint` object.
Endpoint registry becomes `void register(Endpoint endpoint)` rather than `register(Endpoint, HttpHandler)`

Reply:
can also consider having EndpointHttpHandler and putting this metadata into the HttpHandler implementations, and returning a list of those.

Reply:
I am in favor of the proposal of @iamdanfox: `Map<Endpoint, HttpHandler>`. Endpoint is nice to have as a pure metadata tuple and is perfectly suited as a key.

Reply:
I prefer using a tuple type over Map, the hash and equality methods on `Endpoint` shouldn't have any bearing on this functionality.